Use Node 10 for the Docker configuration

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request makes substantial changes to the Docker configuration for this project. Instead of being built on the `node:6` image it instead builds a Node 10 base image on top of the `alpine:3.17` image to use for the app image. This lets the Docker-ization of this project run on an OS that still sees support, bumps from Node 6 to Node 10, and bumps from `npm` v3 to `npm` v7. The last item allows us to use a `package-lock.json` (replacing the `npm-shrinkwrap.json` file) and `npm ci` to fully lock this project's dependencies. This update also allows us to drop the [bower] "hack" as the default registry's TLS certificate is seen as valid. Lastly the `Dockerfile` and `docker-compose.yml` were updated to only bring in the files necessary to function and to only mount the project's JavaScript source code.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Node 10 was the latest version of [Node.js] with which I could successfully perform an `npm install` without failure. Thus it is the most recent version of Node that we can use without investing work into this project's code base, and we can reap any improvements in [Node.js] since the v6 release. Switching the underlying operating system from Debian 9 (Stretch) to Alpine Linux v3.17 both makes the resulting image significantly smaller and puts us on an operating system that is still seeing support. Although Alpine Linux v3.18 has been released I stuck to v3.17 as it is the latest version supported by the [official `node` Docker images](https://github.com/nodejs/docker-node). Updating [Node.js] versions, and by extension `npm` versions, allows us to use a `package-lock.json` file in concert with `npm ci` to perform repeatable Docker image builds. This will hopefully prevent unexpected dependency failures when deploying this configuration in our production environment.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass. This configuration successfully deployed in my development environment and cursory functionality checks were successful. However, I would like to deploy this in the production environment manually and get positive feedback on functionality from the primary users before merging this PR. That deployment would be alongside the current deployment to allow for easily falling back on a known usable configuration.

> [!NOTE]
> I deployed this branch in our production CyHy environment and had the primary users report any feedback. No issues were observed (with some reports of performance gains).
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

[bower]: https://github.com/bower/bower
[Node.js]: https://github.com/nodejs/node

mcdonnnj committed:
Bump from node:6 to node:10 in the Docker configuration

Moving to Node 10 allows us to remove the custom `npm-shrinkwrap.json`
and `.bowerrc` files we needed to successfully build a Docker image on
Node 6.

mcdonnnj committed:
Adjust the Docker configuration

Only bring what is necessary for the app to run into the Docker image
when building. Additionally only mount the source code from the host in
the Docker composition.

mcdonnnj committed:
Change the Docker base image from node:10 to node:10-alpine

This changes the underlying operating system of the image from Debian 9
(Stretch) to Alpine Linux v3.11. Appropriate changes were made to the
Dockerfile, the Makefile configuration and the `entrypoint.sh` script
to accommodate this change. Using an Alpine Linux-based image greatly
reduces the generated image size and streamlines the image by cutting
away unnecessary functionality.

mcdonnnj committed:
Build our own base image

Since Node 10, and the underlying Alpine Linux v3.11, are both long out
of support we can get a bit more support by scratch building a Node 10
base image on top of Alpine Linux 3.17. We then build our Docker
configuration on top of this base image using a multi-stage build.

mcdonnnj committed:
Update the version of npm in the base image

Install the latest version of npm that supports Node 10. This will get
us as up-to-date as we possibly can on such a legacy version of Node.

mcdonnnj committed:
Use a `package-lock.json` file to control dependency versions

This should allow us to lock in our dependencies and produce repeatable
builds to guarantee functionality.

mcdonnnj committed:
Update the Node.js and npm versions referenced in the README

These match the versions used in the updated Docker configuration.

dav3r reviewed (approved):
LGTM! 👍

jsf9k commented on Makefile.docker:
Did you crib this makefile from somewhere else?  The only targets I see being used in the `Dockerfile` are `install` and `gulp`, but there are several others here as well.  I suggest either removing the extra targets or explaining why it's useful to have them around.  (And if this file was cribbed from somewhere else then it would make sense to include a link.)

mcdonnnj replied:
That Makefile is cribbed from the main `Makefile`. It's basically the main one minus any Docker bits and using `/bin/ash` instead of `/bin/bash`.

mcdonnnj replied:
Also the `entrypoint.sh` script uses the `build` and `build-dev` targets depending on which command you provide.

jsf9k replied:
Re "That Makefile is cribbed from the main `Makefile`. It's basically the main one minus any Docker bits and using `/bin/ash` instead of `/bin/bash`." - Could you add a comment to this effect at the top of your cribbed Makefile?

jsf9k replied:
Re "Also the `entrypoint.sh` script uses the `build` and `build-dev` targets depending on which command you provide." - I think adding a comment to this effect to the cribbed Makefile would help the reader follow along, but I'll leave that to your judgement.

mcdonnnj replied:
Please let me know if [74e5a3e](https://github.com/cisagov/ncats-webui/pull/18/commits/74e5a3ef2cdad3e1a24e7bcac8c2fa03410e106f) sufficiently addresses your feedback.

jsf9k replied:
The commit 74e5a3e definitely addresses the first concern I raised [here](https://github.com/cisagov/ncats-webui/pull/18/commits/74e5a3ef2cdad3e1a24e7bcac8c2fa03410e106f#r1205770020), but not the concern I raised [here](https://github.com/cisagov/ncats-webui/pull/18/commits/74e5a3ef2cdad3e1a24e7bcac8c2fa03410e106f#r1205771877).  Did you decide to disregard the latter?

mcdonnnj replied:
I was treating "The other commands are left intact to allow development within a running Docker container as the src/ directory is mapped into the container when using the Docker composition in the project." as a catch-all for the remaining (non-Docker) commands instead of going one-by-one for why a given command was left in the file.

mcdonnnj replied:
@jsf9k Does [d9d4926](https://github.com/cisagov/ncats-webui/pull/18/commits/d9d4926fe25ff585ab3e40aa9490f0682ad35e22) align better with what you were thinking?

jsf9k replied:
Thank you

jsf9k reviewed (changes requested):
I have one concern, but otherwise this looks like your usual strong work.

mcdonnnj committed:
Add explanatory comments to a Makefile

Add comments explaining that Makefile.docker is based on and why it has
the commands it retains from the source Makefile.

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj committed:
Improve comments in the Docker configuration's Makefile

Be more clear about what the commands are used for and add attribution
for any commands that are required by the Docker configuration.

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>
